ruby_atomic.h

```diff
@@ -27,16 +49,107 @@
-{
+/**********************************/
+
+/* Platform-specific implementation (or fallback) */
 #if defined(HAVE_GCC_ATOMIC_BUILTINS)
-    return __atomic_load_n(ptr, __ATOMIC_RELAXED);
+#define DEFINE_ATOMIC_LOAD_EXPLICIT_BODY(ptr, memory_order, type, type_prefix) \
+    __atomic_load_n(ptr, memory_order)
+#elif defined(HAVE_STDATOMIC_H)
+#define DEFINE_ATOMIC_LOAD_EXPLICIT_BODY(ptr, memory_order, type, type_prefix) \
+    atomic_load_explicit((_Atomic volatile type *)ptr, memory_order)
 #else
-    return *ptr;
+#define DEFINE_ATOMIC_LOAD_EXPLICIT_BODY(ptr, memory_order, type, type_prefix) \
+    ((void)memory_order, rbimpl_atomic_##type_prefix##load(ptr))
 #endif
+
+/* Single macro definition for load operations with explicit memory ordering */
+#define DEFINE_ATOMIC_LOAD_EXPLICIT(type_prefix, type) \
+static inline type \
+rbimpl_atomic_##type_prefix##load_explicit(type *ptr, int memory_order) \
+{ \
+    return DEFINE_ATOMIC_LOAD_EXPLICIT_BODY(ptr, memory_order, type, type_prefix); \
 }
-#define ATOMIC_LOAD_RELAXED(var) rbimpl_atomic_load_relaxed(&(var))
+
+/* Generate atomic load function with explicit memory ordering */
+DEFINE_ATOMIC_LOAD_EXPLICIT(, rb_atomic_t)
+DEFINE_ATOMIC_LOAD_EXPLICIT(value_, VALUE)
+DEFINE_ATOMIC_LOAD_EXPLICIT(ptr_, void *)
+
+#undef DEFINE_ATOMIC_LOAD_EXPLICIT
+#undef DEFINE_ATOMIC_LOAD_EXPLICIT_BODY
+
+/* Define the body for two-operand atomic operations based on platform */
+#if defined(HAVE_GCC_ATOMIC_BUILTINS)
+#define ATOMIC_OP_EXPLICIT_BODY(gcc_op, stdatomic_op, ptr, val, memory_order, type, type_prefix) \
+    __atomic_##gcc_op(ptr, val, memory_order)
+#elif defined(HAVE_STDATOMIC_H)
+#define ATOMIC_OP_EXPLICIT_BODY(gcc_op, stdatomic_op, ptr, val, memory_order, type, type_prefix) \
+    atomic_##stdatomic_op##_explicit((_Atomic volatile type *)ptr, val, memory_order)
+#else
+#define ATOMIC_OP_EXPLICIT_BODY(gcc_op, stdatomic_op, ptr, val, memory_order, type, type_prefix) \
+    ((void)memory_order, rbimpl_atomic_##type_prefix##stdatomic_op(ptr, val))
+#endif
+
+/* Generic macro for two-operand atomic operations with explicit memory ordering */
+#define DEFINE_ATOMIC_OP_EXPLICIT(gcc_op, stdatomic_op, type_prefix, type) \
+static inline type \
+rbimpl_atomic_##type_prefix##stdatomic_op##_explicit(volatile type *ptr, type val, int memory_order) \
+{ \
+    return ATOMIC_OP_EXPLICIT_BODY(gcc_op, stdatomic_op, ptr, val, memory_order, type, type_prefix); \
+}
+
+/* Special case for store operations that don't return values */
+#define DEFINE_ATOMIC_STORE_EXPLICIT(gcc_op, stdatomic_op, type_prefix, type) \
+static inline void \
+rbimpl_atomic_##type_prefix##stdatomic_op##_explicit(volatile type *ptr, type val, int memory_order) \
+{ \
+    ATOMIC_OP_EXPLICIT_BODY(gcc_op, stdatomic_op, ptr, val, memory_order, type, type_prefix); \
+}
+
+/* Generate atomic operations with explicit memory ordering */
+DEFINE_ATOMIC_STORE_EXPLICIT(store_n, store, , rb_atomic_t)
+DEFINE_ATOMIC_OP_EXPLICIT(add_fetch, add, , rb_atomic_t)
```

Comment:
I don't think there's an `atomic_add_explicit()` in `<stdatomic.h>` (https://en.cppreference.com/w/c/header/stdatomic.html)